backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,287 @@
+"""Contains derivatives for BatchNorm."""
+from typing import Dict, List, Tuple, Union
+
+from torch import Size, Tensor, einsum
+from torch.nn import BatchNorm1d, BatchNorm2d, BatchNorm3d
+
+from backpack.core.derivatives.basederivatives import BaseParameterDerivatives
+
+
+class BatchNormNdDerivatives(BaseParameterDerivatives):
+    """Derivatives for BatchNorm1d, 2d and 3d.
+
+    If training=False: saved statistics are used.
+    If training=True: statistics of current batch are used.
+
+    Index convention:
+    n: batch axis
+    c: category axis
+    {empty}/l/hw/dhw: dimension axis for 0/1/2/3-dimensions
+
+    Links to PyTorch docs:
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html
+    https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm3d.html
+
+    As a starting point for derivative computation, see these references:
+    https://kevinzakka.github.io/2016/09/14/batch_normalization/
+    https://chrisyeh96.github.io/2017/08/28/deriving-batchnorm-backprop.html
+    """
+
+    def _check_parameters(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> None:
+        if module.affine is False:
+            raise NotImplementedError("Only implemented for affine=True")
+        if module.track_running_stats is False:
+            raise NotImplementedError("Only implemented for track_running_stats=True")
+
+    def hessian_is_zero(self) -> bool:
+        """Whether hessian is zero.
+
+        Returns:
+            whether hessian is zero
+        """
+        return False
+
+    def hessian_is_diagonal(self) -> bool:
+        """Whether hessian is diagonal.
+
+        Returns:
+            whether hessian is diagonal
+        """
+        return False
```

Comment:
Should also depend on the mode. But we currently don't have tests for this function, it is only called by some second-order extensions, which are only tested by the old test suite which complicates adding new settings.

I vote for raising a `NotImplementedError` when `training=False`.

Reply:
Same issue of missing `module` in interface.

Reply:
changed interface